kong/tools/queue.lua

```diff
@@ -0,0 +1,369 @@
+    self.last_used = now()
+    ok, err = self.semaphore:wait(((data_started + self.max_delay) - now()) / 1000)
+    if not ok and err == "timeout" then
+      break
+    elseif ok then
+      entry_count = entry_count + 1
+    else
+      self:log(ERR, "could not wait for semaphore: %s", err)
+      break
+    end
+  end
+
+  local start_time = now()
+  local retry_count = 0
+  while true do
+    self:log(DEBUG, "passing %d entries to handler", entry_count)
+    self.last_used = now()
+    ok, err = self.handler(self, {unpack(self.queue, 1, entry_count)})
+    if ok then
+      self:log(DEBUG, "handler processed %d entries sucessfully", entry_count)
+      break
+    end
+
+    if not err then
+      self:log(ERR, "handler returned falsy value but no error information")
+    end
+
+    if (now() - start_time) > self.max_retry_time then
+      self:log(
+        ERR,
+        "could not send entries, giving up after %d retries.  %d queue entries were lost",
+        retry_count, entry_count)
+      break
+    end
+
+    self:log(WARN, "handler could not process entries: %s", tostring(err))
+    retry_count = retry_count + 1
+    self.last_used = now()
+    ngx.sleep(math.min(self.max_retry_delay, (retry_count * retry_count) * 0.01))
+  end
+
+  for _ = 1, entry_count do
+    self:delete_frontmost_entry()
+  end
+  if self.queue_full then
+    self:log(INFO, 'queue resumed processing')
+    self.queue_full = false
+  end
+end
+
+
+function Queue.get_params(config)
+  local key = config.__key__
+  local queue = unpack({config.queue or {}})
+  if config.retry_count then
+    self:log(ngx.WARN, string.format(
+      "deprecated `retry_count` parameter in plugin %s ignored",
+      key))
+  end
+  if config.queue_size then
+    self:log(ngx.WARN, string.format(
+      "deprecated `queue_size` parameter in plugin %s converted to `queue.batch_max_size`",
+      key))
+    queue.batch_max_size = config.queue_size
+  end
+  if config.flush_timeout then
+    self:log(ngx.WARN, string.format(
+      "deprecated `flush_timeout` parameter in plugin %s converted to `queue.max_delay`",
+      key))
+    queue.max_delay = config.flush_timeout
+  end
+  return queue
+end
+
+
+-------------------------------------------------------------------------------
+-- Drain the queue, for testing purposes.
+-- @param self Queue
+function Queue:drain()
+  self.running = false
+  while #self.queue > 0 do
+    self:process_once(0.01)
+  end
+end
+
+
+-------------------------------------------------------------------------------
+-- Add entry to the queue
+-- @param conf plugin configuration of the plugin instance that caused the item to be queued
+-- @param entry the value included in the queue. It can be any Lua value besides nil.
+-- @return true, or nil and an error message.
+function Queue:add(data)
+  -- FIXME check argument types
+  if data == nil then
+    return nil, "entry must be a non-nil Lua value"
+  end
+
+  if #self.queue >= self.capacity * 0.9 then
```

Comment:
how about adding`local queue_len = #self.queue`?

Reply:
How would that be better?

Reply:
if the queue has too many elements, `#self.queue` will be calculated twice.

Reply:
What do you mean by "calculated".  Is getting the number of entries in an array an expensive operation?